File: src/lib/cloudi_core/src/cloudi_logger.erl
%-*-Mode:erlang;coding:utf-8;tab-width:4;c-basic-offset:4;indent-tabs-mode:()-*-
% ex: set ft=erlang fenc=utf-8 sts=4 ts=4 sw=4 et nomod:
%%%
%%%------------------------------------------------------------------------
%%% @doc
%%% ==CloudI Logger Configuration==
%%% @end
%%%
%%% MIT License
%%%
%%% Copyright (c) 2023 Michael Truog <mjtruog at protonmail dot com>
%%%
%%% Permission is hereby granted, free of charge, to any person obtaining a
%%% copy of this software and associated documentation files (the "Software"),
%%% to deal in the Software without restriction, including without limitation
%%% the rights to use, copy, modify, merge, publish, distribute, sublicense,
%%% and/or sell copies of the Software, and to permit persons to whom the
%%% Software is furnished to do so, subject to the following conditions:
%%%
%%% The above copyright notice and this permission notice shall be included in
%%% all copies or substantial portions of the Software.
%%%
%%% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
%%% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
%%% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
%%% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
%%% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
%%% FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
%%% DEALINGS IN THE SOFTWARE.
%%%
%%% @author Michael Truog <mjtruog at protonmail dot com>
%%% @copyright 2023 Michael Truog
%%% @version 2.0.6 {@date} {@time}
%%%------------------------------------------------------------------------

-module(cloudi_logger).
-author('mjtruog at protonmail dot com').

%% external interface
-export([aspect_stop_on_fatal/10,
         aspect_stop_on_error/10]).

%%-------------------------------------------------------------------------
%% @doc
%% ===Fatal stop function for aspects_log_after logging_set configuration.===
%% Add as {cloudi_logger, aspect_stop_on_fatal}.
%% With the aspect function added to the configuration, CloudI log output
%% will cause CloudI to stop execution after fatal log output.
%% @end
%%-------------------------------------------------------------------------

-spec aspect_stop_on_fatal(Level :: cloudi_service_api:loglevel_on(),
                           Timestamp :: erlang:timestamp(),
                           Node :: node(),
                           Pid :: pid(),
                           FileName :: nonempty_string(),
                           Line :: non_neg_integer(),
                           Function :: atom() | undefined,
                           Arity :: arity() | undefined,
                           MetaData :: list({atom(), any()}) | #{},
                           LogMessage :: iodata()) ->
    ok.

aspect_stop_on_fatal(fatal, _, _, _, _, _, _, _, _, _) ->
    init:stop(1);
aspect_stop_on_fatal(_, _, _, _, _, _, _, _, _, _) ->
    ok.

%%-------------------------------------------------------------------------
%% @doc
%% ===Error stop function for aspects_log_after logging_set configuration.===
%% Add as {cloudi_logger, aspect_stop_on_error}.
%% With the aspect function added to the configuration, CloudI log output
%% will cause CloudI to stop execution after fatal or error log output.
%% @end
%%-------------------------------------------------------------------------

-spec aspect_stop_on_error(Level :: cloudi_service_api:loglevel_on(),
                           Timestamp :: erlang:timestamp(),
                           Node :: node(),
                           Pid :: pid(),
                           FileName :: nonempty_string(),
                           Line :: non_neg_integer(),
                           Function :: atom() | undefined,
                           Arity :: arity() | undefined,
                           MetaData :: list({atom(), any()}) | #{},
                           LogMessage :: iodata()) ->
    ok.

aspect_stop_on_error(Level, _, _, _, _, _, _, _, _, _)
    when Level =:= fatal orelse Level =:= error ->
    init:stop(1);
aspect_stop_on_error(_, _, _, _, _, _, _, _, _, _) ->
    ok.

File: src/lib/cloudi_core/include/CloudILogger.ex
#-*-Mode:ruby;coding:utf-8;tab-width:4;c-basic-offset:4;indent-tabs-mode:()-*-
# ex: set ft=ruby fenc=utf-8 sts=4 ts=4 sw=4 et nomod:
###
###------------------------------------------------------------------------
###
### BSD LICENSE
### 
### Copyright (c) 2014, Michael Truog <mjtruog at gmail dot com>
### All rights reserved.
### 
### Redistribution and use in source and binary forms, with or without
### modification, are permitted provided that the following conditions are met:
### 
###     * Redistributions of source code must retain the above copyright
###       notice, this list of conditions and the following disclaimer.
###     * Redistributions in binary form must reproduce the above copyright
###       notice, this list of conditions and the following disclaimer in
###       the documentation and/or other materials provided with the
###       distribution.
###     * All advertising materials mentioning features or use of this
###       software must display the following acknowledgment:
###         This product includes software developed by Michael Truog
###     * The name of the author may not be used to endorse or promote
###       products derived from this software without specific prior
###       written permission
### 
### THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
### CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
### INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
### OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
### DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
### CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
### SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
### BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
### SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
### INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
### WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
### NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
### OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
### DAMAGE.
###
###------------------------------------------------------------------------

defmodule CloudILogger do

    defmacro log_fatal(format, args) do
        quote do
            :cloudi_core_i_logger_interface.fatal(__MODULE__, __ENV__.line,
                                                  unquote(format),
                                                  unquote(args))
        end
    end

    defmacro log_error(format, args) do
        quote do
            :cloudi_core_i_logger_interface.error(__MODULE__, __ENV__.line,
                                                  unquote(format),
                                                  unquote(args))
        end
    end

    defmacro log_warn(format, args) do
        quote do
            :cloudi_core_i_logger_interface.warn(__MODULE__, __ENV__.line,
                                                 unquote(format),
                                                 unquote(args))
        end
    end

    defmacro log_info(format, args) do
        quote do
            :cloudi_core_i_logger_interface.info(__MODULE__, __ENV__.line,
                                                 unquote(format),
                                                 unquote(args))
        end
    end

    defmacro log_debug(format, args) do
        quote do
            :cloudi_core_i_logger_interface.debug(__MODULE__, __ENV__.line,
                                                  unquote(format),
                                                  unquote(args))
        end
    end

    defmacro log_trace(format, args) do
        quote do
            :cloudi_core_i_logger_interface.trace(__MODULE__, __ENV__.line,
                                                  unquote(format),
                                                  unquote(args))
        end
    end

    defmacro log_fatal_apply(f, a) do
        quote do
            :cloudi_core_i_logger_interface.fatal_apply(unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_error_apply(f, a) do
        quote do
            :cloudi_core_i_logger_interface.error_apply(unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_warn_apply(f, a) do
        quote do
            :cloudi_core_i_logger_interface.warn_apply(unquote(f),
                                                       unquote(a))
        end
    end

    defmacro log_info_apply(f, a) do
        quote do
            :cloudi_core_i_logger_interface.info_apply(unquote(f),
                                                       unquote(a))
        end
    end

    defmacro log_debug_apply(f, a) do
        quote do
            :cloudi_core_i_logger_interface.debug_apply(unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_trace_apply(f, a) do
        quote do
            :cloudi_core_i_logger_interface.trace_apply(unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_fatal_apply(m, f, a) do
        quote do
            :cloudi_core_i_logger_interface.fatal_apply(unquote(m),
                                                        unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_error_apply(m, f, a) do
        quote do
            :cloudi_core_i_logger_interface.error_apply(unquote(m),
                                                        unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_warn_apply(m, f, a) do
        quote do
            :cloudi_core_i_logger_interface.warn_apply(unquote(m),
                                                       unquote(f),
                                                       unquote(a))
        end
    end

    defmacro log_info_apply(m, f, a) do
        quote do
            :cloudi_core_i_logger_interface.info_apply(unquote(m),
                                                       unquote(f),
                                                       unquote(a))
        end
    end

    defmacro log_debug_apply(m, f, a) do
        quote do
            :cloudi_core_i_logger_interface.debug_apply(unquote(m),
                                                        unquote(f),
                                                        unquote(a))
        end
    end

    defmacro log_trace_apply(m, f, a) do
        quote do
            :cloudi_core_i_logger_interface.trace_apply(unquote(m),
                                                        unquote(f),
                                                        unquote(a))
        end
    end

    defmacro function() do
        quote do
            :erlang.element(2,
                :erlang.element(2,
                    :erlang.process_info(:erlang.self(), current_function)))
        end
    end

end
